File: frost-taproot/src/refresh.rs
// Mirrors ref/frost/src/lib/refresh.ts

use k256::Scalar;

use crate::shares::{combine_set, create_shares};
use crate::types::{SecretShare, SecretSharePackage};
use crate::vss::{create_share_coeffs, get_share_commits};
use crate::Error;

/// Generate refresh shares for proactive secret sharing.
/// The polynomial has a zero constant term, so adding these shares to existing
/// shares does not change the underlying secret.
/// Mirrors `gen_refresh_shares` in the TS implementation.
pub fn gen_refresh_shares(
    index: u32,
    threshold: usize,
    share_max: u32,
    secrets: &[[u8; 32]],
) -> Result<SecretSharePackage, Error> {
    // Auxiliary coefficients (threshold - 1 of them, no constant term).
    let sub_coeffs = create_share_coeffs(secrets, threshold - 1);
    // Prepend zero as the constant term so the polynomial evaluates to 0 at x=0.
    let coeffs: Vec<Scalar> = std::iter::once(Scalar::ZERO)
        .chain(sub_coeffs.iter().cloned())
        .collect();

    let shares = create_shares(&coeffs, share_max)?;
    let vss_commits = get_share_commits(&sub_coeffs);

    Ok(SecretSharePackage {
        idx: index,
        vss_commits,
        shares,
    })
}

/// Apply refresh shares to a current share by summing them.
/// Mirrors `refresh_share` in the TS implementation.
pub fn refresh_share(
    refresh_shares: &[SecretShare],
    current_share: &SecretShare,
) -> Result<SecretShare, Error> {
    let all: Vec<SecretShare> = std::iter::once(current_share.clone())
        .chain(refresh_shares.iter().cloned())
        .collect();
    combine_set(&all)
}

File: frost-taproot-go/refresh/refresh.go
// Package refresh provides proactive secret sharing functionality.
package refresh

import (
	"math/big"

	"github.com/frost-taproot/frost-taproot-go/shares"
	"github.com/frost-taproot/frost-taproot-go/types"
	"github.com/frost-taproot/frost-taproot-go/vss"
)

// GenRefreshShares generates refresh shares with zero constant term.
func GenRefreshShares(index uint32, threshold int, shareMax uint32, secrets [][32]byte) (types.SecretSharePackage, error) {
	// Auxiliary coefficients (threshold - 1, no constant term)
	subCoeffs := vss.CreateShareCoeffs(secrets, threshold-1)
	// Prepend zero as constant term
	coeffs := make([]*big.Int, 0, threshold)
	coeffs = append(coeffs, big.NewInt(0))
	for _, c := range subCoeffs {
		coeffs = append(coeffs, c)
	}

	sharesList, err := shares.CreateShares(coeffs, shareMax)
	if err != nil {
		return types.SecretSharePackage{}, err
	}
	vssCommits := vss.GetShareCommits(subCoeffs)

	return types.SecretSharePackage{
		ID:         index,
		Shares:     sharesList,
		VssCommits: vssCommits,
	}, nil
}

// RefreshShare applies refresh shares to a current share.
func RefreshShare(refreshShares []types.SecretShare, currentShare *types.SecretShare) (types.SecretShare, error) {
	all := make([]types.SecretShare, 0, len(refreshShares)+1)
	all = append(all, *currentShare)
	all = append(all, refreshShares...)
	return shares.CombineSet(all)
}
